pkg/controller/clusterconfig/clusterconfig_controller.go

```diff
@@ -0,0 +1,113 @@
+package clusterconfig
+
+import (
+	"context"
+	"log"
+
+	"github.com/pkg/errors"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-network-operator/pkg/apply"
+	"github.com/openshift/cluster-network-operator/pkg/names"
+	"github.com/openshift/cluster-network-operator/pkg/network"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	configv1.Install(mgr.GetScheme())
+	return &ReconcileClusterConfig{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("clusterconfig-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource cluster.openshift.io/v1/Network
+	err = c.Watch(&source.Kind{Type: &configv1.Network{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileClusterConfig{}
+
+// ReconcileClusterConfig reconciles a cluster Network object
+type ReconcileClusterConfig struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile updates the Network Operator CRD based on changes in the Cluster config.
+func (r *ReconcileClusterConfig) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("Reconciling Network %s/%s\n", request.Namespace, request.Name)
+
+	// We won't create more than one network
+	if request.Name != names.CLUSTER_CONFIG {
+		log.Printf("Ignoring Network without default name " + names.CLUSTER_CONFIG)
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch the NetworkConfig instance
+	instance := &configv1.Network{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
```

Comment:
This comment is boilerplate automatically generated by the operator-sdk, FWIW. Since we set the ownerReference, the usual kubernetes garbage-collection rules apply.

However, we don't set the owner reference in this case, because I decided that we don't want to delete the CRD. This is only because the network is special, and deleting the network configuration is likely to be a serious error. In most cases, this is not expected behavior, and I could be convinced to change this.

I'll delete the misleading comment, though.